kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/optimizations/KonanBCEForLoopBodyTransformer.kt

```diff
@@ -0,0 +1,244 @@
+                                ((loopHeader.loopInitStatements[0] as? IrVariable)?.initializer as? IrGetValue)?.symbol)
+                    }
+                    is ProgressionLoopHeader -> analysisResult = analyzeLoopHeader(loopHeader.nestedLoopHeader)
+                }
+        }
+        return analysisResult
+    }
+
+    private fun replaceOperators(expression: IrCall, index: IrExpression, safeIndexVariables: List<IrVariable>): IrExpression {
+        if (index is IrGetValue && index.symbol.owner in safeIndexVariables) {
+            val operatorWithoutBC = expression.dispatchReceiver!!.type.getClass()!!.functions.singleOrNull {
+                if (expression.symbol.owner.name == OperatorNameConventions.SET)
+                    it.name == KonanNameConventions.setWithoutBC
+                else
+                    it.name == KonanNameConventions.getWithoutBC
+            } ?: return expression
+            return IrCallImpl(
+                    expression.startOffset, expression.endOffset, expression.type, operatorWithoutBC.symbol,
+                    typeArgumentsCount = expression.typeArgumentsCount,
+                    valueArgumentsCount = expression.valueArgumentsCount).apply {
+                dispatchReceiver = expression.dispatchReceiver
+                for (argIndex in 0 until expression.valueArgumentsCount) {
+                    putValueArgument(argIndex, expression.getValueArgument(argIndex))
+                }
+            }
+        }
+        return expression
+    }
+
+    override fun visitCall(expression: IrCall): IrExpression {
+        if (!analysisResult.boundsAreSafe || analysisResult.arrayInLoop == null)
+            return expression
+        if (expression.symbol.owner.name != OperatorNameConventions.SET && expression.symbol.owner.name != OperatorNameConventions.GET) {
+            expression.transformChildrenVoid()
+            return expression
+        }
+        if (expression.dispatchReceiver?.type?.isBasicArray() != true ||
+                (expression.dispatchReceiver as? IrGetValue)?.symbol != analysisResult.arrayInLoop)
+            return expression
+        // Analyze arguments of set/get operator.
+        val index = expression.getValueArgument(0)
+        return when(loopHeader) {
+            is ProgressionLoopHeader -> with(loopHeader as ProgressionLoopHeader) {
+                replaceOperators(expression, index!!, listOf(mainLoopVariable, inductionVariable))
+            }
+
+            is WithIndexLoopHeader -> with(loopHeader as WithIndexLoopHeader) {
+                when(nestedLoopHeader) {
+                    is IndexedGetLoopHeader ->
+                        replaceOperators(expression, index!!, listOfNotNull(indexVariable, loopVariableComponents[1]))
+                    is ProgressionLoopHeader ->
+                        replaceOperators(expression, index!!,
+                                listOfNotNull(indexVariable, loopVariableComponents[1], loopVariableComponents[2])
```

Comment:
Can you add a comment explaining why `loopVariableComponents[2]` should be safe. At first I thought it was not, since it is the value in the progression. But then I remembered you are checking that the progression is within the array size bounds.